### ISSUE TITLE
Using the plugins in NodeJS

### ISSUE THREAD

**Issue by florinpop17** [PrismJS/prism#1171]
How can the plugins be used in NodeJS when importing like:
```import Prism from 'prismjs';```

I tried to access the line-numbers plugin, but when I do `Prism.plugins` it's an empty object.

Any ideas?

**Comment by megamindbrian** [comment]
This doesn't work:
```
var prism = require('prismjs');
require('prismjs/plugins/keep-markup/prism-keep-markup.js');
```
In node for me.

**Comment by florinpop17** [comment]
What I did was to copy the entire JS file into my static folder, and use it from there... Not the most elegant thing to do...

**Comment by megamindbrian** [comment]
@florinpop17 Thank you!  Good advice.  I about to go down that path but I've given up for now.  Not sure how document.createRange can even be used from the keep-markup plugin so I'm stuck until deeper investigation.

**Comment by schwastek** [comment]
@megamindbrian both `import 'prismjs'` and `require('prismjs')` should work. They create `Prism` object.

To check if you are able to access the properties of the created `Prism` object, open the DevTools Console and type `Prism` (it's case sensitive). Initially, `Prism.plugins` is an empty object `{}`.

Everytime you require Prism's plugin it injects itself into the `Prism.plugins` object. 
Take a look at `prism-keep-markup.js` file, line 7:
```javascript
Prism.plugins.KeepMarkup = true;
```

In conclusion, this should work for you:
```javascript
import 'prismjs';
import 'prismjs/plugins/keep-markup/prism-keep-markup.js';
```
Maybe you should check your bundler's paths - what directories are being searched when resolving modules and if the `node_modules` directory is being taken into account.

**Comment by megamindbrian** [comment]
Thank you for circling back. I'll try that but I am trying to run from
node. Basically trying to combine prism with node diff to get an even
prettier output than GitHub gives you in code compare.

**Comment by Golmote** [comment]
@florinpop17 @megamindbrian At the moment, most plugins won't work on Node because they use the DOM (Line numbers and Keep markup are one of those).

Are you trying to use those with some virtual DOM node modules or something?

**Comment by megamindbrian** [comment]
Haven't tried it with virtual DOM.  Good idea though.

**Comment by Golmote** [comment]
@mAAdhaTTah Most plugin don't work in Node indeed. From a quick look:

- **Autolinker**: Should work if the `before-highlight` hook is run manually.
- Autoloader: Does not work and returns early. Requires DOM.
- Command line: Does not work and returns early. Requires DOM.
- Copy to clipboard: Does not work and returns early. Requires DOM.
- **Custom class**: Should work.
- **Data-URI highlight**: Should work if the `before-highlight` hook is run manually.
- File Highlight: Does not work and returns early. Requires DOM.
- **Highlight keywords**: Should work.
- JSONP highlight: Does not work and returns early. Requires DOM.
- Keep Markup: Does not work and returns early. Requires DOM.
- Line Highlight: Does not work and returns early. Requires DOM.
- Line Numbers: Does not work and returns early. Requires DOM.
- **Normalize Whitespace**: Should work if the `before-sanity-check` hook is run manually.
- Previewers: Does not work and returns early. Requires DOM.
- **Remove initial line feed**: Deprecated, but should work if the `before-sanity-check` hook is run manually.
- **Show invisibles**: Should work if the `before-highlight` hook is run manually.
- Show language: Does not work and returns early. Requires DOM.
- Toolbar: Does not work and returns early. Requires DOM.
- Unescaped markup: Does not work and returns early. Requires DOM.
- **WPD**: Should work.

I left this issue open, because I know this is a common expectation from our users. Yet I think you are right, most of the plugins that do not work currently probably won't ever be compatible with Node in the future.

Regarding a `loadPlugins` function, is it something that would be useful? They usually don't have any dependency...

**Comment by mAAdhaTTah** [comment]
> Regarding a `loadPlugins` function, is it something that would be useful? They usually don't have any dependency...

The main benefit would be not needing to know the paths. I dunno if it's _that_ beneficial, but it does at least provide some symmetry with the way `loadLanguages` works.

**Comment by benwiley4000** [comment]
This is a major bummer especially when using something like Spectacle/react-live which relies on Node prism for the client side. Any thoughts on improving this? :slightly_smiling_face:

**Comment by mAAdhaTTah** [comment]
Right now the suggestion is to use the babel plugin: https://github.com/mAAdhaTTah/babel-plugin-prismjs/ – there are definitely still details to iron out if you're providing a library using Prism, as it suggests the consumer of the library needs to run babel over their dependencies (or at least that library), which may or may not be easy to do.

**Comment by benwiley4000** [comment]
I may be incorrect but I don't think this will work for the Node API (which react-live uses for rendering server and client side) because the plugins (line numbers, line highlight) rely on DOM attributes being present, and prism node doesn't operate on DOM elements.

**Comment by mAAdhaTTah** [comment]
That's correct. May not be worth it / possible to highlight server-side in that case.

**Comment by benwiley4000** [comment]
Right. The issue is that most React based libraries including this one use the node variant of a library even for the client side because it fits much better into the way React renders. So more and more consumers of popular libraries like spectacle are going to be unable to take full advantage of Prismjs's client side capabilities, which is a shame. :/

**Comment by mAAdhaTTah** [comment]
They can highlight on `componentDidUpdate`, so it doesn't fire on the server. They also should not be using `loadLanguages` in any code intended to be used on the client.

**Comment by lunelson** [comment]
AFAICT currently it relies on the `global` object to register the plugins on the Prism object. This works for me:
```js
global.Prism = require('prismjs');
require('prismjs/plugins/normalize-whitespace/prism-normalize-whitespace');
require('prismjs/plugins/data-uri-highlight/prism-data-uri-highlight');

Object.keys(Prism.plugins); // [ 'NormalizeWhitespace', 'dataURIHighlight' ]
```

**Comment by tig** [comment]
I was really expecting Prism's line number support to work from node. Found this issue and I'm quite bummed.

I'm using node.js to generate static html files from code, formatting them with prismjs, so there's no DOM to access.

**Comment by xcatliu** [comment]
> Autolinker: Should work if the before-highlight hook is run manually

**Comment by RunDevelopment** [comment]
You can run any hook using `Prism.hooks.run('<hook name>', env);`. Insert any hook name (e.g. `before-highlight`). 
`env` is tricky but for Autolinker, [we only need a grammar](https://github.com/PrismJS/prism/blob/9c7bc820787ff30cd33b25d00f438ea60873c478/plugins/autolinker/prism-autolinker.js#L52).

```js
const code = `var foo = 0;`;
const language = 'javascript';

const grammar = Prism.languages[language];
Prism.hooks.run('before-highlight', { grammar });
const highlighted = Prism.highlight(code, grammar, language);
```

Please keep in mind that hooks are mostly internal at this point in time, so this code my break at any time.

TBH, Autolinker, DataURI URL, and Show Invisibles should be implemented in a way that works only with the hooks from `Prism.highlight`, so you don't have to manually call any hooks.

**Comment by asm0dey** [comment]
Is it possible to implement line-highlight in browser-independent way? For example highlightjs (which is much less powerful_ has it.

**Comment by peterbe** [comment]
> Is it possible to implement line-highlight in browser-independent way?

I think you mean NodeJS :)
Yes, I'm very much in favor of this plugin working in Node. 

Line-highlighting and line-numbering are the only plugins I genuinely need for Node.

**Comment by judocode** [comment]
Hopefully this helps someone as I didn't find an answer in any of these comments. I am using Prism with React via NPM and was having some difficulty with getting line numbers to work. It seems line-numbers is relying on the `complete` hook, so I was able to get it to work by manually calling it after the component was mounted. This is what I came up with that seems to work (yes I use class components still and css styles are coming from Tailwind):

```typescript
import * as React from 'react'
import Prism from 'prismjs'
import DOMPurify from 'dompurify'
import 'prismjs/plugins/line-numbers/prism-line-numbers.js'
const language = 'javascript'

const grammar = Prism.languages[language]

interface IProps {
  code: string
}

class CodeComponent extends React.Component<IProps, {}> {
  codeRef: React.RefObject<HTMLDivElement>

  constructor(props: IProps) {
    super(props)

    this.codeRef = React.createRef()
  }

  componentDidMount() {
    // @ts-ignore
    Prism.hooks.run('complete', {
      grammar,
      element: this.codeRef.current,
      code: this.props.code,
    })
  }

  render(): React.ReactNode {
    const code = Prism.highlight(this.props.code, grammar, language)

    return (
      <pre className="pr-6 py-4 pl-12">
        <code
          ref={this.codeRef}
          className="language-js block line-numbers relative"
          dangerouslySetInnerHTML={{
            __html: DOMPurify.sanitize(
              code.trim().replaceAll('\n', '<br />'),
            ),
          }}
        />
      </pre>
    )
  }
}

export default CodeComponent
```

**Comment by TrebledJ** [comment]
Used the same strat as quilicicf, but used [domino](https://github.com/fgnass/domino) instead of jsdom for speed improvements and zero additional dependencies. Got command-line, line-numbers, and toolbar to work for my Eleventy site.

Writeup: https://trebledj.me/posts/prism-plugins-in-node/
